sqld/src/query.rs

```diff
@@ -268,53 +265,83 @@ impl ToSql for Value {
     }
 }
 
+#[derive(Debug)]
+pub enum Params {
+    Named(HashMap<String, Value>),
+    Positional(Vec<Value>),
+}
+
+impl Params {}
+
 impl Params {
     pub fn empty() -> Self {
-        Self { params: Vec::new() }
+        Self::Positional(Vec::new())
     }
 
-    pub fn new(params: Vec<(Option<String>, Value)>) -> Self {
-        Self { params }
+    pub fn new_named(values: HashMap<String, Value>) -> Self {
+        Self::Named(values)
     }
 
-    pub fn push(&mut self, name: Option<String>, value: Value) {
-        self.params.push((name, value));
+    pub fn new_positional(values: Vec<Value>) -> Self {
+        Self::Positional(values)
     }
 
-    fn get_name(&self, k: &str) -> Option<&Value> {
-        // strip prefix ('$', '?', ..)
-        let mut chars = k.chars();
-        chars.next();
-        let stripped = chars.as_str();
-
-        if let Ok(index) = stripped.parse::<usize>() {
-            return self.get_pos(index);
+    pub fn get_pos(&self, pos: usize) -> Option<&Value> {
+        assert!(pos > 0);
+        match self {
+            Params::Named(_) => None,
+            Params::Positional(params) => params.get(pos - 1),
         }
+    }
 
-        self.params.iter().find_map(|(name, val)| match name {
-            Some(name) if name == stripped => Some(val),
-            _ => None,
-        })
+    pub fn get_named(&self, name: &str) -> Option<&Value> {
+        match self {
+            Params::Named(params) => params.get(name),
```

Comment:
Yes it is because the prefix is a part of the variable name, in other words :name and $name refer to two different variables

Reply:
because that's how sqlite does it, most libraries seem to follow it:
 - https://docs.rs/rusqlite/latest/rusqlite/macro.named_params.html
 - https://sql.js.org/#/
 - https://learn.microsoft.com/en-us/dotnet/standard/data/sqlite/parameters

and so forth... With the notable exception of Python: https://docs.python.org/3/library/sqlite3.html#how-to-guides

But this is problematic as well, at least for error handling, because now SQLite reports that more parameters are being bound than there actually are if you ignore the prefix, so to report an error you'd need to keep track of what parameters have been used so far, and which are left to be bound. Perfectly doable, but is it worth 1) the hassle and 2) diverging from what seems conventional at this point?